R/gwas_mappings.R

```diff
@@ -0,0 +1,199 @@
+    colnames(Y) <- c("strain", "trait")
+
+    # Sort kinship matrix to to be in the same order as phenotyped strains
+    kinship_sorted = kinship[sort(row.names(kinship)),sort(colnames(kinship))]
+    # Prune kinship matrix to only include phenotyped individuals
+    K = kinship_sorted[row.names(kinship_sorted) %in% Y$strain,
+                       colnames(kinship_sorted) %in% Y$strain]
+    # Sort kinship matrix to to be in the same order as phenotyped strains
+    markers = genotype %>%
+        tidyr::unite(marker, CHROM, POS, remove = FALSE ) %>%
+        dplyr::select(marker, CHROM, POS, dplyr::everything(), -REF, -ALT)
+
+    markers_sorted <- data.frame(markers[,1:3],
+                                 markers[, sort(names(
+                                     markers[names(markers)%in%Y$strain]))])
+
+    # ID markers that have a MAF outside the user-defined range
+    keepMarkers <- data.frame(
+        MAF = apply(markers_sorted[,4:ncol(markers_sorted)],
+                    MARGIN = 1,
+                    FUN = function(x){
+                        x[x==-1] <- 0
+                        return(sum(x, na.rm = T)/length(x))}
+        )) %>%
+        dplyr::mutate(marker = markers_sorted$marker) %>%
+        dplyr::filter(MAF >= min.MAF) %>%
+        dplyr::filter(MAF <= 1 - min.MAF)
+
+    # Remove markers identified to be out of MAF range
+    M <- markers_sorted %>%
+        dplyr::filter(marker %in% keepMarkers$marker)
+
+    if (map_by_chrom) {
+        # need 6 different kinship matrices
+        # chrom2:x, map on one
+        # chrom1:5, map on x, etc
+        # then subtract blups from phenotype and map on
+        # chromosome that you are not accounting for relationship
+
+        # I think these are two ways to correct phenotype for relatedness
+        yfit <- rrBLUP::mixed.solve(y = Y$trait, K = K)
+        kfit <- rrBLUP::kin.blup(data = data.frame(Y),
+                                 geno = "strain",
+                                 pheno = "trait",
+                                 K = K)
+
+        # then get blup residuals
+        y_blup <- Y$trait - yfit$u
+        # then map outside the EMMA framework, similar to linkage mapping
+        # (−n(ln(1−r2)/2ln(10)))
+        # repeat for all chromosomes
+    } else {
+        # Perform mapping
+        gwa_results = rrBLUP::GWAS(pheno = data.frame(Y),
+                                   geno = data.frame(M),
+                                   K = K,
+                                   n.PC = 0,
+                                   min.MAF = min.MAF,
+                                   n.core = mapping_cores,
+                                   P3D = P3D,
+                                   plot = FALSE)
+    }
+
+    # Process mapping results
+    gwa_results_pr <- gwa_results %>%
+        dplyr::rename(log10p = trait) %>%
+        dplyr::filter(log10p != 0) %>%
+        dplyr::mutate(BF = -log10(FDR_threshold/n()), # set bonferroni threshold
```

Comment:
FYI Variables can also be placed above dplyr verbs without interupting the pipe. Not necessary to change, just making you aware if you were not.